pymongo/message.py

```diff
@@ -878,6 +923,305 @@ def _do_batched_op_msg(
+        abridged_keys.append("comment")
+    if command.get("let"):
+        abridged_keys.append("let")
+    command_abridged = {key: command[key] for key in abridged_keys}
+    command_len_abridged = len(_dict_to_bson(command_abridged, False, opts))
+
+    # When OP_MSG is used unacknowledged we have to check command
+    # document size client-side or applications won't be notified.
+    if not ack:
+        _check_doc_size_limits("bulkWrite", command_abridged, max_bson_size + _COMMAND_OVERHEAD)
+
+    # Maximum combined size of the ops and nsInfo document sequences.
+    max_doc_sequences_bytes = max_message_size - (_OP_MSG_OVERHEAD + command_len_abridged)
+
+    ns_info = {}
+    to_send_ops: list[Mapping[str, Any]] = []
+    to_send_ns: list[Mapping[str, int]] = []
+    total_ops_length = 0
+    total_ns_length = 0
+    idx = 0
+
+    for real_op_type, op_doc in operations:
+        op_type = real_op_type
+        # Check insert/replace document size if unacknowledged.
+        if real_op_type == "insert":
+            if not ack:
+                _check_doc_size_limits(real_op_type, op_doc["document"], max_bson_size)
+        if real_op_type == "replace":
+            op_type = "update"
+            if not ack:
+                _check_doc_size_limits(real_op_type, op_doc["updateMods"], max_bson_size)
+
+        ns_doc_to_send = None
+        ns_length = 0
+        namespace = op_doc[op_type]
+        if namespace not in ns_info:
+            ns_doc_to_send = {"ns": namespace}
+            new_ns_index = len(to_send_ns)
+            ns_info[namespace] = new_ns_index
+
+        # First entry in the operation doc has the operation type as its
+        # key and the index of its namespace within ns_info as its value.
+        op_doc_to_send = copy.deepcopy(op_doc)
+        op_doc_to_send[op_type] = ns_info[namespace]  # type: ignore[index]
+
+        # Encode current operation doc and, if newly added, namespace doc.
+        op_length = len(_dict_to_bson(op_doc_to_send, False, opts))
+        if ns_doc_to_send:
+            ns_length = len(_dict_to_bson(ns_doc_to_send, False, opts))
+
+        # Check operation document size if unacknowledged.
+        if not ack:
+            _check_doc_size_limits(op_type, op_doc_to_send, max_bson_size + _COMMAND_OVERHEAD)
+
+        new_message_size = total_ops_length + total_ns_length + op_length + ns_length
+        # We have enough data, return this batch.
+        if new_message_size > max_doc_sequences_bytes:
+            break
+        to_send_ops.append(op_doc_to_send)
+        total_ops_length += op_length
+        if ns_doc_to_send:
+            to_send_ns.append(ns_doc_to_send)
+            total_ns_length += ns_length
+        idx += 1
+        # We have enough documents, return this batch.
+        if idx == max_write_batch_size:
+            break
+
+    # Construct the entire OP_MSG.
+    length = _client_construct_op_msg(command_doc, to_send_ops, to_send_ns, ack, opts, buf)
```

Comment:
command_doc is bytes here but _client_construct_op_msg accepts a `command: Mapping[str, Any]`. Which type is correct? Shouldn't the mypy test be failing on that?

Reply:
I think the types are correct since `_client_batched_op_msg_impl()` accepts a `command: Mapping[str, Any]` but `_client_construct_op_msg()` accepts a `command_doc: bytes`. I'll change them both to accept a `command: Mapping[str, Any]` so that it's less confusing!

Reply:
Oh I see, I misread then. No need to change the existing approach then.